GH-1902: Fix misleading system contract gas charges

mpapierski opened the pull request:
Closes #1902 

This PR introduces an improved system contract charge method that charges depending on the immediate caller being a system contract or user.

The problem was that `auction::ADD_BID` entry point calls `mint::CREATE_PURSE,` which misled gas cost reporting.

This change also fills the gap where the call stack was inconsistent across the system - for example, in genesis, the installer didn't set it correctly, and the improved charging method wouldn't be reliable. This PR also ensures that invalid (the first frame that's not a session) or empty call stack is a programming error.

Commit:
Add regression test

Commit:
Fix the cascading cost issue.

This change will ensure that if a system contract calls another system
contract then the charge isn't accrued, but otherwise such calls will be
charged if started from a session or non-system stored ontract.

luis-casper commented on execution_engine/src/core/runtime/auction_internal.rs:
```diff
@@ -195,7 +195,12 @@ where
 
         let gas_counter = self.gas_counter();
-        let call_stack = self.call_stack().clone();
+        let mut call_stack = self.call_stack().clone();
```
Why do we need to clone call stacks? Wouldn't it be more natural to have the call functions push or pop from them on demand?
This is, of course, out of the scope of this PR.

mpapierski replied:
This is because `call_host_mint` assumes the call stack is already updated. We do push to call stack before creating a new Runtime object. I.e. in `Executor::exec` it already runs with an initial call stack (session + variant for stored contracts). In `Runtime::execute_contract` we're pushing a new element before `call_host_mint` is called. Auction internals are only outlier and without this change implemented checks wouldn't work consistently

luis-casper commented on execution_engine/src/core/runtime/mod.rs:
```diff
@@ -2893,6 +2904,22 @@ where
+    fn get_system_contract_stack_frame(&mut self, name: &str) -> Result<CallStackElement, Error> {
+        let contract_hash = self.context.get_system_contract(name)?;
+        let key = Key::from(contract_hash);
+        let contract = match self.context.read_gs(&key)? {
```
We shouldn't have to read from global state to get the hash of system contracts because they are part of the node software.

darthsiroftardis replied:
@luis-casper We write the system contract hashes to global state, and the node retrieves it by reading it from there. Do you have an example where the node has an alternate source?

luis-casper replied:
We have `SystemContractRegistry`, for instance.

mpapierski replied:
System contract hashes are, in fact, part of node software, but we need to be able to query them off-chain, so we've moved them into a global state. (previously, we've had a separate database for "ProtocolData," which was complicating the design of fast sync

luis-casper replied:
Sure, but this code is part of the node software, so we could avoid that read, right?

luis-casper replied:
So the bottom line is that this isn't really possible since we need the package hash also.

luis-casper commented on execution_engine/src/core/runtime/mod.rs:
```diff
@@ -3639,6 +3666,50 @@ where
+    ///
+    /// For cases where call stack is only the session code, then this method returns `true` if the
+    /// caller is system, or `false` otherwise.
+    fn is_system_immediate_caller(&self) -> Result<bool, Error> {
```
I'm concerned with this code for a few reasons:
* First, doesn't `get_immediate_caller` already consult the call stack? Why is there a need to do it twice?
* Second, despite the name of the method being `is_system_immediate_caller`, it looks at the first entry of the stack, not the last, as does `get_immediate_caller`.

mpapierski replied:
> First, doesn't get_immediate_caller already consult the call stack? Why is there a need to do it twice?

Get immediate caller checks only the one element before the end of the stack. This method verifies that the caller is a system one. 

> Second, despite the name of the method being is_system_immediate_caller, it looks at the first entry of the stack, not the last, as does get_immediate_caller.

Fair point on naming. I re-checked the necessity to consult the first element in case of a lack of an immediate caller. The immediate caller is assumed to be present when `call_host_*` methods are calling this, and there's no further check needed for the first entry. Addressed in 7d5b9a79d59498d9b59f3853f8afa10b7ec3ec75

luis-casper commented on execution_engine_testing/tests/src/test/regression/gh_1902.rs:
```diff
@@ -0,0 +1,158 @@
+fn setup() -> InMemoryWasmTestBuilder {
+    let mut builder = InMemoryWasmTestBuilder::default();
+    builder.run_genesis(&*DEFAULT_RUN_GENESIS_REQUEST);
+    let id: Option<u64> = None;
```
Why not just put `Option::<u64>::None` below directly?

mpapierski replied:
Consistency - all of the testing code uses this pattern, so I went with it.

luis-casper commented on execution_engine_testing/tests/src/test/regression/gh_1902.rs:
```diff
@@ -0,0 +1,158 @@
+fn setup() -> InMemoryWasmTestBuilder {
+    let mut builder = InMemoryWasmTestBuilder::default();
+    builder.run_genesis(&*DEFAULT_RUN_GENESIS_REQUEST);
+    let id: Option<u64> = None;
```
Why not just put `Option::<u64>::None` below directly?

Commit:
Assume presence of immediate caller in call site.

This change does not require check of the front element considering that
call stack is consistent.

Fraser999 commented on execution_engine/src/core/runtime/mod.rs:
```diff
@@ -987,6 +987,17 @@ where
+            call_stack.first().unwrap().contract_hash().is_none(),
+            "First element should always represent a Session call"
+        );
+
```
We should log errors here for each of these `debug_assert!` cases.

Fraser999 commented on execution_engine/src/core/runtime/mod.rs:
```diff
@@ -1032,6 +1043,9 @@ where
     where
         T: Into<Gas>,
     {
+        if self.is_system_immediate_caller()? {
```
Please update doc comment.

Fraser999 commented on execution_engine_testing/tests/src/test/regression/gh_1902.rs:
```diff
@@ -0,0 +1,158 @@
+    },
+    Gas, PublicKey, RuntimeArgs, SecretKey, U512,
+};
+use once_cell::sync::Lazy;
```
Should be in its own group.

Fraser999 commented on execution_engine_testing/tests/src/test/regression/gh_1902.rs:
```diff
@@ -0,0 +1,158 @@
+        balance_after,
+        expected,
+        "before and after should match; off by: {}",
+        expected - balance_after
```
Could overflow?

Fraser999 reviewed (approved):
Approved pending nitpicks being handled.

Commit:
Address @Fraser999's review feedback.

Commit:
Add extra comment

Commit:
Merge branch 'dev' into gh-1902-fix-create-purse-charge

mpapierski commented:
bors r+

Commit:
Fix clippy

mpapierski commented:
bors r+
